java/ql/src/experimental/Security/CWE/CWE-548/InsecureDirectoryConfig.ql

```diff
@@ -0,0 +1,43 @@
+/**
+ * @name Inappropriately exposed directories and files yielding sensitive information like source code and credentials to attackers.
+ * @description A directory listing provides an attacker with the complete index of all the resources located inside of the complete web directory.
+ * @kind problem
+ * @id java/dir-listing
+ * @tags security
+ *       external/cwe-548
+ */
+
+import java
+import semmle.code.xml.WebXML
+
+/**
+ * The default `<servlet-class>` element in a `web.xml` file.
+ */
+private class DefaultTomcatServlet extends WebServletClass {
+  DefaultTomcatServlet() {
+    this.getTextValue() = "org.apache.catalina.servlets.DefaultServlet" //Default servlet of Tomcat and other servlet containers derived from Tomcat like Glassfish
+  }
+}
+
+/**
+ * The `<init-param>` element in a `web.xml` file, nested under a `<servlet>` element controlling directory listing.
+ */
+class DirectoryListingInitParam extends WebXMLElement {
+  DirectoryListingInitParam() {
+    getName() = "init-param" and
+    getAChild("param-name").getTextValue() = "listings" and
+    exists(WebServlet servlet |
+      getParent() = servlet and servlet.getAChild("servlet-class") instanceof DefaultTomcatServlet
+    )
+  }
+
+  /**
+   * Check the `<param-value>` element (true - enabled, false - disabled)
+   */
+  predicate isListingEnabled() { getAChild("param-value").getTextValue().toLowerCase() = "true" }
+}
+
+from DirectoryListingInitParam initp
+where initp.isListingEnabled()
+select initp, "Directory listing should be disabled to mitigate filename and path disclosure"
+
```

Comment:
Now there's a newline too many.  The newly added commit has appended two newline chars to the file.

Reply:
Sorry @aschackmull . The extra newline character must be added when I copied the auto-formatted code to the staging repository for committing. It seems that the correct way is to transfer the whole file through SFTP.

I noticed the file is one character shorter so the issue shall have been addressed.

Reply:
Somehow the check of "Code scanning - action / CodeQL-Build (pull_request)" still failed. Does the issue with the ending newline character still exist?  

Thanks,
@luchua-bc

Reply:
Sorry @aschackmull . I've removed the extra newline character.